docs/builder/ref.md

```diff
@@ -0,0 +1,290 @@
+
+To reference a component you did not [store a handle][hndl], you can do so later with the `Builder().get_component()` method.
+
+```python
+prog = cb.Builder()
+prog.component("my_component")
+# a few lines later 
+my_component = prog.get_component("my_component")
+```
+
+### Defining component inputs and outputs
+
+Components can be given input and output ports. All you have to do is specify the name of the port and its size.
+
+```python
+my_component.input("my_input", 32)
+my_component.output("my_output", 32)
+```
+
+You can access the input and output ports of a component within the definition of a component using the standard `this.port` syntax.
+
+```python
+def add_my_component(prog):
+    my_component = prog.component("my_component")
+
+    this = my_component.this()
+    with my_component.group("my_group")
+      # some other assignments
+        this.out = sum.out
+```
+
+### Component Cells
+
+### Multi-Component Designs
+
+Calyx supports [multi-component designs][multi]. The [top-level example](ref.md#top-level-program-structure) demonstrates how to construct multi-component designs using the library.
+
+#### Defining Common Calyx Cells
+
+Here's a snippet of code that adds a few common kinds of cells to a component:
+
+```python
+my_component = prog.component("my_component")
+
+# Registers: reg(name, bitwidth)
+my_component.reg("my_reg", 32)
+
+# Constants: const(name, bitwidth, value)
+my_component.const("my_reg", 32, 42)
+
+# Adders: 
+my_component.add()
+
+# Subtractors: 
+
+# Comparators: [gt|lt|eq|neq|ge|le](name, size, signed=False)
+my_component.gt("my_gt", 32)
+# A signed lt comparison
+my_component.lt("my_lt", 32, signed=True)
+my_component.eq("my_eq", 32)
+my_component.neq("my_neq", 32)
+my_component.ge("my_ge", 32)
+my_component.le("my_le", 32)
+
+# 1-D memory: 
+# mem_d1(name, bitwidth, len, idx_size, is_external=False, is_ref=False)
+my_component.mem_d1("my_mem", 32, 4, 32)
+# An external memory
+my_component.mem_d1("my_mem", 32, 4, 32, is_external=True)
+# A memory by reference
+my_component.mem_d1("my_mem", 32, 4, 32, is_ref=True)
+
+
+
+```
+
+If you're curious, you can read more about [external memories][ext] or [memories by reference][ref].
+
+#### Retrieving Cells
+
+## Wires
+
+### Groups
+
+#### Defining Groups
+
+[Groups][grp] are defined using the `group()` method for a component.
+
+```python
+my_component = prog.component("my_component")
+
+with my_component.group("my_group") as my_group:
+    # group assignments here
+```
+
+#### Defining Combinational Groups
+
+[Combinational groups][comb] are defined similarly to groups, except with the `comb_group` method.
+
+```python
+my_component = prog.component("my_component")
+
+with my_component.comb_group("my_comb_group"):
+    # group assignments here
+```
+
+#### Retrieving Groups
+
+If you didn't [store a handle](ref.md#creating-handles-to-components) to your group when you initialized it, you can do so later with the `Builder().get_group()` method.
```

Comment:
Worth noting that this is how you look up _both_ combinational and standard groups